Intersections_2/include/CGAL/Intersections_2/internal/Triangle_2_Triangle_2_intersection_impl.h

```diff
@@ -308,35 +319,68 @@ Triangle_2_Triangle_2_pair<K>::intersection_point() const
+    typename K::FT res(0);
+    std::size_t length = ps.size();
+    for(std::size_t i = 0; i<length; ++i)
+      res += (ps[(i+1)%length].x() - ps[i].x())*(ps[(i+1)%length].y()+ps[i].y());
+
+    return res > 0;
+  }
+};
+
+template <typename K>
+struct Is_cw<K, Tag_true>
+{
+  bool operator()(const std::vector<typename K::Point_2>& ps)
+  {
+    return CGAL::right_turn(ps[0], ps[1], ps[2]);
+  }
+};
 
 template <class K>
 typename CGAL::Intersection_traits
 <K, typename K::Triangle_2, typename K::Triangle_2>::result_type
 intersection(const typename K::Triangle_2 &tr1, 
-	     const typename K::Triangle_2 &tr2,
-	     const K&)
+             const typename K::Triangle_2 &tr2,
+             const K&)
 {
-    typedef Triangle_2_Triangle_2_pair<K> is_t;
-    is_t ispair(&tr1, &tr2);
-    switch (ispair.intersection_type()) {
+  typedef Triangle_2_Triangle_2_pair<K> is_t;
+  is_t ispair(&tr1, &tr2);
+  switch (ispair.intersection_type())
+  {
     case is_t::NO_INTERSECTION:
     default:
-        return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>();
+      return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>();
     case is_t::POINT:
-        return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_point());
+      return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_point());
     case is_t::SEGMENT:
-        return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_segment());
+      return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_segment());
     case is_t::TRIANGLE:
-        return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_triangle());
-    case is_t::POLYGON: {
-        typedef std::vector<typename K::Point_2> Container;
-        Container points(ispair.vertex_count());
-        for (int i =0; i < ispair.vertex_count(); i++) {
-            points[i] = ispair.vertex(i);
-        }
-        return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(points);
-    }
+      return intersection_return<typename K::Intersect_2, typename K::Triangle_2, typename K::Triangle_2>(ispair.intersection_triangle());
+    case is_t::POLYGON:
+    {
+      typedef std::vector<typename K::Point_2> Container;
+      Container points(ispair.vertex_count());
+      for (int i =0; i < ispair.vertex_count(); i++)
+        points[i] = ispair.vertex(i);
+
+      if(Is_cw<K, typename Algebraic_structure_traits<typename K::FT>::Is_exact>()(points))
+      {
+        std::size_t length = points.size();
+
+        for(std::size_t i = 0; i< length/2; ++i)
+          std::swap(points[i], points[length-i-1]);
```

Comment:
just call `std::reverse`?

Reply:
No, it would start at length/2-1 instead of length-1